src/library/plugin/fp/pluginLoader.ts

```diff
@@ -151,51 +142,92 @@ const initializePlugin = (
-      ),
-      TE.mapLeft(
-        (err) => `error while initializing plugin ${pluginName}: ${err}`
-      )
+export const initializePlugins =
+  (
+    // the root logger of the hierarchy
+    rootLogger: Logger
+  ) =>
+  (cfg: Configuration): TaskEither<string, PluginRuntimeEnv> => {
+    const resolvePluginModulePath =
+      (name: string, optionalPath?: string) => () => {
+        const dir =
+          cfg.pluginDir === undefined
+            ? process.cwd()
+            : path.resolve(cfg.pluginDir);
+
+        return optionalPath === undefined
+          ? path.join(dir, name)
+          : path.resolve(optionalPath);
+      };
+
+    const resolveModulePathAndInitializePlugin =
+      (initLogger: Logger, pluginsLogger: Logger) =>
+      (
+        pluginName: string,
+        pluginCfg: {
+          path?: string | undefined;
+          config?: unknown;
+        }
+      ): TaskEither<string, RuntimePluginInstance> =>
+        pipe(
+          TE.Do,
+          TE.bind('modulePath', () =>
+            TE.fromIO(resolvePluginModulePath(pluginName, pluginCfg.path))
+          ),
+          TE.let(
+            // The configuration object passed to the `initialize` function of the plugin factory
+            'pluginConfig',
+            () => pluginCfg.config ?? {}
+          ),
+          TE.tapIO(({ modulePath, pluginConfig }) => () => {
+            initLogger.info(`loading plugin ${pluginName}`);
+            initLogger.debug({
+              pluginName,
+              modulePath,
+              pluginConfig
+            });
+          }),
+          TE.bind('pluginLogger', () =>
+            TE.fromIO(() =>
+              pluginsLogger.getSubLogger({
+                name: `${pluginName}`
+              })
+            )
+          ),
+          TE.chain(({ modulePath, pluginConfig, pluginLogger }) =>
+            initializePlugin(modulePath, pluginConfig, pluginLogger)
+          ),
+          TE.mapLeft(
+            // TODO: more structural error?
+            (err) => `error while initializing plugin ${pluginName}: ${err}`
+          ),
+          TE.tapError((err) =>
+            TE.fromIO(() =>
+              initLogger.error(`unable to initialize plugin ${pluginName}`, err)
+            )
+          )
+        );
+
+    const Applicative = TE.getApplicativeTaskValidation(
+      T.ApplyPar,
+      pipe(Str.Semigroup, S.intercalate(', '))
     );
 
-  const Applicative = TE.getApplicativeTaskValidation(
-    T.ApplyPar,
-    pipe(Str.Semigroup, S.intercalate(', '))
-  );
+    const mkLogger = (name: string) =>
+      TE.fromIO(() => rootLogger.getSubLogger({ name }));
 
-  return R.traverseWithIndex(Applicative)(resolveModulePathAndInitializePlugin)(
-    cfg.plugins
-  );
-};
+    return pipe(
+      TE.Do,
+      TE.bind('initLogger', () => mkLogger('pluginLoader')),
+      TE.bind('pluginsLogger', () => mkLogger('activePlugins')),
+      TE.bind('runtimeLogger', () => mkLogger('pluginRuntime')),
```

Comment:
Also if we decided to go with subloggers it might a good idea to allow plugins to be configured with different loggers per-plugin

Reply:
They can already do that just by modifying the `settings` field of the logger object.

Reply:
Thanks